### ISSUE TITLE
Test that the APK caching works with concurrent fetches.

### ISSUE THREAD

**Issue by mattmoor** [chainguard-dev/go-apk#63]
I had to fix an issue with this [here](https://github.com/chainguard-dev/go-apk/pull/62) already, but we should make sure that fetching a substantial number of packages concurrently don't lead to random concurrency failures, or the cache becomes more trouble than it's worth.

**Comment by mattmoor** [comment]
I'm pretty sure that this code still has bugs wrt `etag` handling since we cannot write two files to the filesystem atomically, which I believe only affects the `APKINDEX.tar.gz` path.  I have hit this a handful of times with highly concurrent builds driven by terraform:
![image](https://github.com/chainguard-dev/go-apk/assets/2442466/85c243a2-f816-453c-a77f-4cc35022fe72)

Looking at the GCLB logs, we haven't served a `404` in the past 7 days, so I'm fairly certain it's another race in the way we are caching.

**Comment by mattmoor** [comment]
@deitch Josh just hit this too.  It generally seems to go away with retry (it did not for josh), but we should really shake any remaining concurrency bugs out of the caching stuff.

We should probably also make sure we have good testing for the path where `etag` changes.

**Comment by mattmoor** [comment]
It looks like the index-not-found issue is one of the most prevalent issues we are seeing with the TF-based builds, which are highly concurrent.  Here is a link to another (private) run where we hit this a fair number of times (my hypothesis is as the index is being changed out from under the build, but that's based purely on intuition): https://github.com/chainguard-images/images-private/actions/runs/5326615430/jobs/9648888060?pr=411

**Comment by deitch** [comment]
Working under the assumption that these 404s are not coming from upstream, as you said above @mattmoor, then it stands to reason that they are coming from within go-apk. Tracing the error message through, it almost definitely has to be from within the http-client that implements the read-through (is it really read-through? I guess) cache. That is [here](https://github.com/chainguard-dev/go-apk/blob/main/pkg/apk/cache.go).

That error only appears in two places. In both cases, it successfully found the cache file via `os.Stat(cacheFile)` [here](https://github.com/chainguard-dev/go-apk/blob/main/pkg/apk/cache.go#L61) first/

* it failed to read the etag file, but etag was not required, so it just tried to read the cache file and failed [here](https://github.com/chainguard-dev/go-apk/blob/main/pkg/apk/cache.go#L73-L75)
* when it tries to open the file to read, it gets an error [here](https://github.com/chainguard-dev/go-apk/blob/main/pkg/apk/cache.go#L96-L99)

In both cases, it successfully `os.Stat()` the cache file but could not read it. That could be a permissions issue, or it could be "process/thread A opened it for writing but did not write it yet, so process/thread B's `os.Stat()` returned, but `os.Open()` failed."

Some experiments should prove it, but also some logging. I will throw in a branch with some absurd logging to help trace it.

**Comment by deitch** [comment]
Ah, it also might be if thread A statted and found, then thread B found an etag mismatch so it deleted.

**Comment by deitch** [comment]
@mattmoor can you try using [this branch](https://github.com/chainguard-dev/go-apk/tree/absurd-cache-logging)? Run it, we should get lots of nice logging.

**Comment by deitch** [comment]
I am fairly convinced it is in that area, the question is, what kind of race condition handling can we do for multiple processes accessing the same files in the same cache directory?

**Comment by mattmoor** [comment]
A potentially dumb suggestion, but building around ideas inspired by content-addressability... for files where we care about the `etag` we could instead write them as: `{etag value}.etag`, and once that's written update `APKINDEX.tar.gz` to symlink to the version of the file observed with that etag (if we even need it).

Unfortunately this means that we leak `APKINDEX.tar.gz` files over time, but hopefully it avoids the race. 🤞 

cc @jonjohnsonjr

**Comment by deitch** [comment]
I think that partially addresses it. In general, we have a potentially multi-process (not only multi-thread) writing to the same cache. We need some locking mechanism. I feel like I am recreating two-phase commit.

**Comment by jonjohnsonjr** [comment]
> I feel like I am recreating two-phase commit.

No need! Use this: https://github.com/google/renameio

**Comment by mattmoor** [comment]
I think that for etag things, we should just use a single file named for the etag, which we write using the retrieveAndSave method.

If we think it's necessary we can change the current os.Rename to use renameio, but I think that shifting the etag-based path to use etag-addressing is probably our best short-term solution here.

**Comment by mattmoor** [comment]
I am playing around with some changes where the etag-path uses `<etag>.etag` for the cache file name, guided by an unconditional `HEAD` to determine the etag here: https://github.com/chainguard-dev/go-apk/compare/main...mattmoor:etag-addressing

I am going to try to reproduce the `APKINDEX.tar.gz` problem locally by doing a `terraform plan` in a loop with blowing away my APK cache to see how reliably I can reproduce this before/after pulling in this change.

**Comment by mattmoor** [comment]
Ok, so one trick I was able to do to make this reproduce immediately (4x in a row) was to run this alongside of a `terraform plan`:
```
watch rm -rf ~/Library/Caches/dev.chainguard.go-apk/https%3A%2F%2Fpackages.wolfi.dev%2Fos/
```

Doing the same with my change above has now worked 4x in a row.

One quick is that for some reason the filename is coming through with quotes, but it seems to work.

**Comment by mattmoor** [comment]
I haven't taken a look at the tests to figure out how to update them to support this.  We're hitting this error a fair amount, so I'd like to land something for this by EoW.

**Comment by deitch** [comment]
> No need! Use this: https://github.com/google/renameio

Sure, anything that makes the problem go away.

I suspect we might be able to recreate this with running several goroutines in parallel. I am going to try that later today.

**Comment by deitch** [comment]
> @mattmoor can you try using [this branch](https://github.com/chainguard-dev/go-apk/tree/absurd-cache-logging)? Run it, we should get lots of nice logging.

Did you manage to do this, and can you post the output? It would be good to confirm the issue.

**Comment by deitch** [comment]
> No need! Use this: https://github.com/google/renameio

**Comment by mattmoor** [comment]
> Did you manage to do this, and can you post the output?

No, but the `terraform plan` with parallel `rm -rf <cache>` repro is trivial and had a 100% reproduction rate, and if you want to use `images-private` it only needs the `APKINDEX.tar.gz` which is public.

You can use a local terraform provider build with: https://github.com/chainguard-dev/terraform-provider-apko/blob/c8d38d20ff310968bae838e905b60cb4443ab104/.github/workflows/mega-module.yaml#L47-L57 to pull in patches

**Comment by mattmoor** [comment]
I think that's backwards.  Things not breaking is >>> things going fast.  If I have to rerun 5-10 times, then it doesn't matter if it's 2x faster.

**Comment by jonjohnsonjr** [comment]
Oh, that's not an ordered list.

**Comment by mattmoor** [comment]
LOL ok, then I totally agree 🤣

**Comment by deitch** [comment]
> Do atomic renames based on etags or checksums for correctness

Actually, we already _do_ atomic renames. The etag files are tiny and are a single write, and in any case do not return a 404, and the cache files are larger and are "write to temp and rename" (courtesy of Matt IIRC). Other than upstream, the _only_ place that can return a 404 is in the cache if it successfully stats the cache file but then cannot open it. That is it.

I have tried various parallel tests, none of them triggers it. I cannot figure out what code path is supposed to cause it.

**Comment by deitch** [comment]
@mattmoor wrote:

> Ok, so one trick I was able to do to make this reproduce immediately (4x in a row) was to run this alongside of a terraform plan:
>
> watch rm -rf ~/Library/Caches/dev.chainguard.go-apk/https%3A%2F%2Fpackages.wolfi.dev%2Fos/
> Doing the same with my change above has now worked 4x in a row.

I can see how that would cause it. If the timing is just right, it can stat the file, then before it reads it, your `watch` removes it. But why is that a valid test case? I am sure I can break lots of things by going around the software's controls to the OS.

**Comment by mattmoor** [comment]
> so remove it all and download a new one

This is why I tested with `rm -rf` of the cache, it's brute force, but exhibited the behavior we are seeing very clearly.

**Comment by deitch** [comment]
I know. But it's so artificial, that I've got no way of knowing if it's something else triggering it when you see it in regular usage, so even if we fix it for that, I don't know if it fixes your issue. 

It also makes it difficult to create a normal regular test for it. I can simulate `rm -rf` in a goroutine, but it's not a normal path.

**Comment by mattmoor** [comment]
I left this open because despite adding testing for the `APKINDEX.tar.gz` thing along with my fix, we still need coverage of the non-etag-required (APK) path.
